test/common/buffer/owned_impl_test.cc

```diff
@@ -1080,6 +1080,83 @@ void TestBufferMove(uint64_t buffer1_length, uint64_t buffer2_length,
   EXPECT_EQ(0, buffer2.length());
 }
 
+TEST_F(OwnedImplTest, CopyOutToSlicesTests) {
+  std::string data = "Hello, World!";
+  Buffer::OwnedImpl buffer;
+  buffer.prepend(data);
+
+  EXPECT_EQ(data.size(), buffer.length());
+  EXPECT_EQ(data, buffer.toString());
+
+  {
+    Buffer::OwnedImpl buf;
+    auto reservation = buf.reserveSingleSlice(1024);
+    auto slice = reservation.slice();
+    EXPECT_EQ(data.size(), buffer.copyOutToSlices(100, &slice, 1));
+    reservation.commit(data.size());
+    EXPECT_EQ(data, buffer.toString());
+  }
+
+  {
+    Buffer::OwnedImpl buf;
+    auto reservation = buf.reserveSingleSlice(5);
+    auto slice = reservation.slice();
+    EXPECT_EQ(5, buffer.copyOutToSlices(100, &slice, 1));
+    reservation.commit(5);
+    EXPECT_EQ("Hello", buf.toString());
+  }
+
+  {
+    Buffer::OwnedImpl buf;
+    auto reservation = buf.reserveForRead();
+    EXPECT_EQ(5, buffer.copyOutToSlices(5, reservation.slices(), reservation.numSlices()));
+    reservation.commit(5);
+    EXPECT_EQ("Hello", buf.toString());
+  }
+
+  {
+    Buffer::OwnedImpl buf;
+    auto reservation = buf.reserveForRead();
+    EXPECT_EQ(data.size(),
+              buffer.copyOutToSlices(100, reservation.slices(), reservation.numSlices()));
+    reservation.commit(data.size());
+    EXPECT_EQ(data, buffer.toString());
+  }
+  // Test the destination buffer has smaller slice than the source buffer.
+  {
+    Buffer::OwnedImpl buf;
+    buf.appendSliceForTest("aa", 2);
+    buf.appendSliceForTest("aa", 2);
+    buf.appendSliceForTest("aa", 2);
+    buf.appendSliceForTest("aa", 2);
+    buf.appendSliceForTest("aa", 2);
+    buf.appendSliceForTest("aa", 2);
+    buf.appendSliceForTest("a", 1);
+    auto reservation = buf.reserveForRead();
+    EXPECT_EQ(data.size(),
+              buffer.copyOutToSlices(100, reservation.slices(), reservation.numSlices()));
+    reservation.commit(data.size());
+    EXPECT_EQ(data, buffer.toString());
```

Comment:
I think this comparison should have been on buf, not buffer.

Reply:
Yes, thanks. That is totally explained why I can write this code and passed all the tests by one-shot....